cp23/hpi-fun2/roomcount/executables/generator.cpp

```diff
@@ -0,0 +1,151 @@
+
+void intervalGraph(vector<int> &intervalStarts, int intervalLength, vector<pair<int, int>> &edges) {
+  vector<int> startIndices(intervalStarts.size());
+  iota(startIndices.begin(), startIndices.end(), 0);
+  sort(startIndices.begin(), startIndices.end(), [&intervalStarts](int idx1, int idx2) {
+    return intervalStarts[idx1] < intervalStarts[idx2];
+  });
+  for (int i = 0; i < startIndices.size(); i++) {
+    int end = intervalStarts[startIndices[i]] + intervalLength;
+    int iterator = i + 1;
+    while (iterator < startIndices.size() && intervalStarts[startIndices[iterator]] < end) {
+      edges.push_back({startIndices[i] + 1, startIndices[iterator] + 1});
+      iterator++;
+    }
+  }
+}
+
+int randomTestcase(int intervalLength, int n, int maxStart) {
+  vector<int> intervalStarts;
+  for (int i = 0; i < n; i++) {
+    intervalStarts.push_back(rnd.next(0, maxStart));
+  }
+  vector<pair<int, int>> edges;
+  intervalGraph(intervalStarts, intervalLength, edges);
+  shuffle(edges.begin(), edges.end());
+
+  cout << n << " " << edges.size() << endl;
+  for (auto& edge : edges) {
+    cout << edge.first << " " << edge.second << endl;
+  }
+  return edges.size();
+}
+
+template <class F>
+void testcase(string name, string desc, F f) {
+    ofstream desc_file(name + ".desc");
+    desc_file << desc;
+    string in_file = name + ".in";
+    freopen(in_file.c_str(), "w", stdout);
+    int m = f();
+    desc_file << ", m = " << m;
+}
+
+void predefined(string name, string desc, string_view content) {
+    ofstream desc_file(name + ".desc");
+    desc_file << desc;
+    string in_file = name + ".in";
+    freopen(in_file.c_str(), "w", stdout);
+    cout << content;
+}
+
+void sample(int num, string_view content) {
+    auto num_str = toString(num);
+    predefined("sample" + num_str, "Sample #" + num_str, content);
+}
+
+int main(int argc, char* argv[]) {
+    registerGen(argc, argv, 1);
+    rnd.setSeed(7889991971352209491ll);
+
+    sample(1, SAMPLE1);
+    sample(2, SAMPLE2);
+
+    for (int i = 0; i < 10; i++) {
+      int n = rnd.next(100, 1000);
+      int intervalLength = rnd.next(5, 150);
+      testcase("random_testcase_" + to_string(i),
+               "random testcase with interval length = " + to_string(intervalLength) + ", n = " + to_string(n),
+               [&n, &intervalLength]() -> int { return randomTestcase(intervalLength, n, n * n / 100); });
+    }
+
+    int maxSize = 100000;
+    testcase("max_size_length",
+             "max length testcase with n = " + to_string(maxSize),
+             [&maxSize]() -> int { return maxLengthTestcase(maxSize); });
+
+    testcase("max_size_width",
+             "max width testcase with n = " + to_string(maxSize),
+             [&maxSize]() -> int { return maxWidthTestcase(maxSize); });
+
+    testcase("max_size_random",
+             "max size random testcase with n = " + to_string(maxSize / 10),
+             [&maxSize]() -> int { return randomTestcase(100, maxSize / 10, maxSize); });
```

Comment:
Is that necessary? Since the seed is fixed, the generated testcase should always be the same, so while theoretically, the generated testcase might have too many edges, the testcase that is actually used has less than `maxSize` edges.

Reply:
It would be a lot cleaner. Also note that every tiny change of a function that uses randomness that is called before changes this testcase.

Reply:
Number of edges is now ensured to be below `maxSize`